shared/src/ledger/treasury/mod.rs

```diff
@@ -0,0 +1,180 @@
+                    } else {
+                        return false;
+                    };
+                    let is_max_funds_transfer_key =
+                        treasury_storage::get_max_transferable_fund_key();
+                    let balance_key = token::balance_key(&nam(), &ADDRESS);
+                    let max_transfer_amount =
+                        self.ctx.read_pre(&is_max_funds_transfer_key);
+                    let pre_balance = self.ctx.read_pre(&balance_key);
+                    let post_balance = self.ctx.read_post(&balance_key);
+                    if addr.ne(&ADDRESS) {
+                        return true;
+                    }
+                    match (max_transfer_amount, pre_balance, post_balance) {
+                        (
+                            Ok(max_transfer_amount),
+                            Ok(pre_balance),
+                            Ok(post_balance),
+                        ) => {
+                            match (
+                                max_transfer_amount,
+                                pre_balance,
+                                post_balance,
+                            ) {
+                                (
+                                    Some(max_transfer_amount),
+                                    Some(pre_balance),
+                                    Some(post_balance),
+                                ) => {
+                                    let max_transfer_amount =
+                                        token::Amount::try_from_slice(
+                                            &max_transfer_amount[..],
+                                        )
+                                        .ok();
+                                    let pre_balance =
+                                        token::Amount::try_from_slice(
+                                            &pre_balance[..],
+                                        )
+                                        .ok();
+                                    let post_balance =
+                                        token::Amount::try_from_slice(
+                                            &post_balance[..],
+                                        )
+                                        .ok();
+                                    match (
+                                        max_transfer_amount,
+                                        pre_balance,
+                                        post_balance,
+                                    ) {
+                                        (
+                                            Some(max_transfer_amount),
+                                            Some(pre_balance),
+                                            Some(post_balance),
+                                        ) => {
+                                            post_balance > pre_balance
+                                                || (pre_balance - post_balance
```

Comment:
I think this will panic with underflow when the treasure receives tokens. We can use e.g. `checked_sub` - maybe we should implement it on `token::Amount`:

```rust
impl Amount {
  pub fn checked_sub(self, rhs: Self) -> Option<Self> {
    let micro = self.micro.checked_sub(rhs.micro)?;
    Some(Self { micro })
  }
}
```